streams/test-utils/src/test/java/org/apache/kafka/streams/MockProcessorContextTest.java

```diff
@@ -0,0 +1,392 @@
+import org.apache.kafka.streams.processor.Processor;
+import org.apache.kafka.streams.processor.ProcessorContext;
+import org.apache.kafka.streams.processor.PunctuationType;
+import org.apache.kafka.streams.processor.Punctuator;
+import org.apache.kafka.streams.processor.TaskId;
+import org.apache.kafka.streams.processor.To;
+import org.apache.kafka.streams.state.KeyValueStore;
+import org.apache.kafka.streams.state.internals.InMemoryKeyValueStore;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.File;
+import java.util.Iterator;
+import java.util.Properties;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertFalse;
+import static org.junit.Assert.assertTrue;
+
+public class MockProcessorContextTest {
+    /**
+     * Behavioral test demonstrating the use of the context for capturing forwarded values
+     */
+    @Test
+    public void shouldCaptureOutputRecords() {
+        final AbstractProcessor<String, Long> processor = new AbstractProcessor<String, Long>() {
+            @Override
+            public void process(final String key, final Long value) {
+                this.context().forward(key + value, key.length() + value);
+            }
+        };
+
+        final MockProcessorContext context = new MockProcessorContext();
+
+        processor.init(context);
+
+        processor.process("foo", 5L);
+        processor.process("barbaz", 50L);
+
+        final Iterator<KeyValue> forwarded = context.forwarded().iterator();
+        assertEquals(forwarded.next(), new KeyValue<>("foo5", 8L));
+        assertEquals(forwarded.next(), new KeyValue<>("barbaz50", 56L));
+        Assert.assertFalse(forwarded.hasNext());
+
+        context.resetForwards();
+
+        assertEquals(context.forwarded().size(), 0);
+    }
+
+    /**
+     * Behavioral test demonstrating the use of the context for capturing forwarded values to specific children
+     */
+    @Test
+    public void shouldCaptureRecordsOutputToChildByName() {
+        final AbstractProcessor<String, Long> processor = new AbstractProcessor<String, Long>() {
+            private int count = 0;
+
+            @Override
+            public void process(final String key, final Long value) {
+                final To child = count % 2 == 0 ? To.child("george") : To.child("pete");
+                this.context().forward(key + value, key.length() + value, child);
+                count++;
+            }
+        };
+
+        final MockProcessorContext context = new MockProcessorContext();
+
+        processor.init(context);
+
+        processor.process("foo", 5L);
+        processor.process("barbaz", 50L);
+
+        {
+            final Iterator<KeyValue> forwarded = context.forwarded().iterator();
+            assertEquals(forwarded.next(), new KeyValue<>("foo5", 8L));
+            assertEquals(forwarded.next(), new KeyValue<>("barbaz50", 56L));
+            Assert.assertFalse(forwarded.hasNext());
```

Comment:
since this testing forwarding to child nodes by name and we have this test above, do we need to have it again here?

Reply:
We don't strictly need it, but these tests are partially intended to demonstrate example usage, so I wrote it in "scenario style".